bootScripts/NodeThreadWorkerBootScript/DSUCodeFileCacheHandler.js

```diff
@@ -0,0 +1,129 @@
+const fs = require("fs");
+const path = require("path");
+
+const readFileAsync = $$.promisify(fs.readFile.bind(fs));
+const writeFileAsync = $$.promisify(fs.writeFile.bind(this));
+const mkdirAsync = $$.promisify(fs.mkdir.bind(this));
+
+const pathExistsAsync = async (path) => {
+    try {
+        await $$.promisify(fs.access.bind(fs))(path);
+        return true;
+    } catch (error) {
+        return false;
+    }
+};
+
+function isFileInsideFolderStructure(file) {
+    return file.indexOf("/") !== -1;
+}
+
+class DSUCodeFileCacheHandler {
+    constructor(dsu, cacheFolderBasePath) {
+        this.dsu = dsu;
+        this.cacheFolderBasePath = cacheFolderBasePath;
+    }
+
+    async constructCache() {
+        const openDSU = require("opendsu");
+        const keySSISpace = openDSU.loadAPI("keyssi");
+        const resolver = openDSU.loadApi("resolver");
+
+        const mountedDSUs = await $$.promisify(this.dsu.listMountedDossiers)("/");
+
+        let codeFolderName = openDSU.constants.CODE_FOLDER;
+        if (codeFolderName[0] === "/") {
+            codeFolderName = codeFolderName.substring(1);
+        }
+        const codeMount = mountedDSUs.find((mount) => mount.path === codeFolderName);
+        const codeDSU = await $$.promisify(resolver.loadDSU)(codeMount.identifier);
+        const codeFiles = await $$.promisify(codeDSU.listFiles)("/");
+
+        // commenting out due to getAllVersions failing on Android
+        // const anchoringApi = openDSU.loadAPI("anchoring");
+        // const codeMountKeySSI = keySSISpace.parse(codeMount.identifier);
+        // const versions = await $$.promisify(anchoringApi.getAllVersions)(codeMountKeySSI);
```

Comment:
The *anchoringApi.getAllVersions* method is failing when running Android emulator on Windows machine.

This is caused by Invalid format for keySSI 6Cn1oQzTMcRLGs5M4HcZmQvEq2VhbM2ATkhw24VuGtNDQGjNeacBhHesXWPPgqkR7HCQSFCBNtG48rpCVGB6heYcz1pHsog42ifJmxxNVyAeGs6MGmQ8Ecw3LLKMqXqxrDJEpD3xtKCk33TQsnXC6dY3Q4tBoeTQHzRifzhexsdfcuE17jggdPB1TpXa9tz5SyBsUB6PdM145UwsH1GyME5LYPwu1jm\r',

Issue comes from **apihub\components\anchoring\strategies\fsx\filePersistence.js**, where **versions** array contains the hashlink with an '\r' at the end, which caused the keyssi parse method to not work as expected.

```
 self.getAllVersions = function (anchorId, callback) {
        const filePath = path.join(anchoringFolder, anchorId);
        fs.readFile(filePath, (err, fileHashes) => {
            if (err) {
                return callback(new Error(`Failed to read file <${filePath}>`));
            }
            const fileContent = fileHashes.toString().trimEnd();
            const versions = fileContent ? fileContent.split(endOfLine) : [];
            callback(undefined, versions);
        });
    }
```



```
9013/eu.pharmaledger.epi E/nodejs: /data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:32028
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:         throw createOpenDSUErrorWrapper(`Invalid format for keySSI ${identifier}`, e);
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:         ^
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs: Error: Cannot read property 'toString' of undefined
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     at ErrorWrapper (/data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:43382:22)
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     at createOpenDSUErrorWrapper (/data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:43409:12)
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     at KeySSIFactory.create (/data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:32028:15)
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     at Object.parse (/data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:44413:26)
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     at /data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:33959:64
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     at Array.map (<anonymous>)
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     at /data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:33959:42
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     at /data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:1020:13
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     at FSReqCallback.readFileAfterClose [as oncomplete] (internal/fs/read_file_context.js:63:3) {
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:   originalMessage: "Cannot read property 'toString' of undefined",
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:   previousError: TypeError: Cannot read property 'toString' of undefined
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     at Object.target.autoLoad (/data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:32213:63)
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     at KeySSIFactory.create (/data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:32026:16)
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     at Object.parse (/data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:44413:26)
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:       at /data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:33959:64
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     at Array.map (<anonymous>)
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:       at /data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:33959:42
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:       at /data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:1020:13
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:       at FSReqCallback.readFileAfterClose [as oncomplete] (internal/fs/read_file_context.js:63:3),
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:   debug_message: 'Invalid format for keySSI 6Cn1oQzTMcRLGs5M4HcZmQvEq2VhbM2ATkhw24VuGtNDQGjNeacBhHesXWPPgqkR7HCQSFCBNtG48rpCVGB6heYcz1pHsog42ifJmxxNVyAeGs6MGmQ8Ecw3LLKMqXqxrDJEpD3xtKCk33TQsnXC6dY3Q4tBoeTQHzRifzhexsdfcuE17jggdPB1TpXa9tz5SyBsUB6PdM145UwsH1GyME5LYPwu1jm\r',
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:   debug_stack: "TypeError: Cannot read property 'toString' of undefined\n" +
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     '    at Object.target.autoLoad (/data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:32213:63)\n' +
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     '    at KeySSIFactory.create (/data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:32026:16)\n' +
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     '    at Object.parse (/data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:44413:26)\n' +
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     '    at /data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:33959:64\n' +
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     '    at Array.map (<anonymous>)\n' +
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     '    at /data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:33959:42\n' +
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     '    at /data/user/0/eu.pharmaledger.epi/files/nodejs-project/pskWebServer.js:1020:13\n' +
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs:     '    at FSReqCallback.readFileAfterClose [as oncomplete] (internal/fs/read_file_context.js:63:3)'
2022-07-26 21:44:02.170 8917-9013/eu.pharmaledger.epi E/nodejs: }
```